Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp

```diff
@@ -0,0 +1,168 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebGPUSurfaceCocoa.h"
+
+#include "PlatformCALayer.h"
+#include "pal/graphics/WebGPU/WebGPUCanvasConfiguration.h"
+#include "pal/graphics/WebGPU/WebGPUDevice.h"
+#include "pal/graphics/WebGPU/WebGPUTextureDescriptor.h"
+
+namespace WebCore {
+
+static bool validateSurfaceCreation(IntSize, const PAL::WebGPU::CanvasConfiguration& configuration)
+{
+    // FIXME: Support RGBA8 or RGBA16Float if possible.
+    if (configuration.format != PAL::WebGPU::TextureFormat::Bgra8unorm) {
+        // FIXME: this generates a device validation error.
+        return false;
+    }
+
+    // FIXME: Support view formats
+    if (!configuration.viewFormats.isEmpty())
+        return false;
+
+    return true;
+}
+
+RefPtr<WebGPUSurfaceCocoa> WebGPUSurfaceCocoa::create(IntSize size, PAL::WebGPU::CanvasConfiguration configuration)
+{
+    if (!validateSurfaceCreation(size, configuration))
+        return nullptr;
+
+    return adoptRef(*new WebGPUSurfaceCocoa(WTFMove(size), WTFMove(configuration)));
+}
+
+WebGPUSurfaceCocoa::WebGPUSurfaceCocoa(IntSize size, PAL::WebGPU::CanvasConfiguration configuration)
+    : m_size(WTFMove(size))
+    , m_configuration(WTFMove(configuration))
+{
+    ASSERT(validateSurfaceCreation(size, configuration));
+}
+
+void WebGPUSurfaceCocoa::prepareToDelegateDisplay(PlatformCALayer&)
+{
+}
+
+void WebGPUSurfaceCocoa::display(PlatformCALayer& layer)
+{
+    if (m_frontBuffer)
+        layer.setContents(m_frontBuffer->surface());
+}
+
+GraphicsLayer::CompositingCoordinatesOrientation WebGPUSurfaceCocoa::orientation() const
+{
+    return GraphicsLayer::CompositingCoordinatesOrientation::TopDown;
+}
+
+void WebGPUSurfaceCocoa::present()
+{
+    std::swap(m_frontBuffer, m_backBuffer);
+    m_backBufferTexture = nullptr;
+    std::swap(m_backBuffer, m_secondaryBackBuffer);
+}
+
+bool WebGPUSurfaceCocoa::isValidBuffer(const IOSurface* buffer)
+{
+    if (!buffer)
+        return false;
+
+    if (buffer->size() != m_size)
+        return false;
+
+    if (buffer->isInUse())
+        return false;
+
+    return true;
+}
+
+void WebGPUSurfaceCocoa::ensureValidBackBuffer()
+{
+    if (isValidBuffer(m_backBuffer.get()))
+        return;
+
+    // The back buffer is not valid, so we should not reuse it and the back buffer texture.
+    // Destroy the back buffer texture here.
+    if (m_backBufferTexture) {
+        m_backBufferTexture->destroy();
+        m_backBufferTexture = nullptr;
+    }
+
+    // Then try to recycle the secondary back buffer as the new back buffer.
+    m_backBuffer = WTFMove(m_secondaryBackBuffer);
+
+    if (isValidBuffer(m_backBuffer.get()))
+        return;
+
+    // Can't use the secondary back buffer, so destroy it and create a new one.
+    m_backBuffer = IOSurface::create(nullptr, m_size, colorSpace());
```

Comment:
I don't think this will work with IOKit sandboxing. This code is expected to run in the Web process, right? I don't think the web process is allowed to create IOSurfaces.

Reply:
This should be likely created by the GPU process, right?